src/BTCPayServer.Lightning.LNDhub/LndHubClient.cs

```diff
@@ -0,0 +1,192 @@
+                : null;
+        }
+
+        public async Task<InvoiceData[]> GetInvoices(CancellationToken cancellation)
+        {
+            return await Get<InvoiceData[]>("getuserinvoices", cancellation);
+        }
+
+        public async Task<TransactionData[]> GetTransactions(CancellationToken cancellation)
+        {
+            return await Get<TransactionData[]>("gettxs", cancellation);
+        }
+
+        public async Task<InvoiceData> CreateInvoice(CreateInvoiceParams req, CancellationToken cancellation)
+        {
+            var payload = new CreateInvoiceRequest
+            {
+                Amount = (long)req.Amount.ToUnit(LightMoneyUnit.Satoshi),
+                Memo = req.Description ?? string.Empty,
+                DescriptionHash = req.DescriptionHash
+            };
+
+            return await Post<CreateInvoiceRequest, InvoiceData>("addinvoice", payload, cancellation);
+        }
+
+        public async Task<PaymentData> Pay(string bolt11, PayInvoiceParams payParams, CancellationToken cancellation)
+        {
+            var payload = new PayInvoiceRequest
+            {
+                PaymentRequest = bolt11,
+            };
+
+            if (payParams?.Amount != null)
+                payload.Amount = (long)payParams.Amount.ToUnit(LightMoneyUnit.Satoshi);
+
+            return await Post<PayInvoiceRequest, PaymentData>("payinvoice", payload, cancellation);
+        }
+
+        private async Task<TResponse> Get<TResponse>(string path, CancellationToken cancellation)
+        {
+            return await Send<EmptyRequestModel, TResponse>(HttpMethod.Get, path, null, cancellation);
+        }
+
+        private async Task<TResponse> Post<TRequest, TResponse>(string path, TRequest payload, CancellationToken cancellation)
+        {
+            return await Send<TRequest, TResponse>(HttpMethod.Post, path, payload, cancellation);
+        }
+
+        private async Task<TResponse> Send<TRequest, TResponse>(HttpMethod method, string path, TRequest payload, CancellationToken cancellation)
+        {
+            HttpContent content = null;
+            if (payload != null)
+            {
+                var payloadJson = JsonConvert.SerializeObject(payload);
+                content = new StringContent(payloadJson, Encoding.UTF8, "application/json");
+            }
+
+            var req = new HttpRequestMessage
+            {
+                RequestUri = new Uri($"{_baseUri}{path}"),
+                Method = method,
+                Content = content
+            };
+            req.Headers.Accept.Clear();
+            req.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
+            req.Headers.Add("User-Agent", "BTCPayServer.Lightning.LndHubClient");
+
+            if (path != "auth" && path != "create")
+            {
+                var accessToken = await GetAccessToken(cancellation);
```

Comment:
you're generating the access token EVERY call? You should try and cache as much as possible (and uswe the refresh token when needed)

Reply:
How can we achieve this? We are stateless here – maybe I'm overlooking something.

Reply:
It is stateless, but you can at least save the access token inside the instance for its lifetime